hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/NearCachedClientCacheProxy.java

```diff
@@ -231,39 +231,49 @@ protected void onPutIfAbsentAsyncInternal(K key, V value, Data keyData, Data val
     @Override
     protected void getAllInternal(Set<? extends K> keys, Collection<Data> dataKeys, ExpiryPolicy expiryPolicy,
                                   List<Object> resultingKeyValuePairs, long startNanos) {
-        Map<Object, Data> keyMap = null;
         if (serializeKeys) {
-            keyMap = toDataKeys(keys, dataKeys);
+            toDataKeysWithReservations(keys, dataKeys, null, null);
         }
-        Collection<?> ncKeys = serializeKeys ? dataKeys : new ArrayList<Object>(keys);
+        Collection<?> ncKeys = serializeKeys ? dataKeys : new ArrayList<K>(keys);
```

Comment:
Not directly related to this PR but: `LinkedList` can fit better here since its `remove` complexity is O(1) instead of O(n) of `ArrayList `